Complete Issues #9 Implement KPR-F1 feature (add new keeper)

Solykk opened the pull request:
connect to #9

Solykk committed:
#9 Created entity Keeper and exception UnsupportedKeeperException
implemented KPR-F1

Solykk committed:
#9 Added Setter for dismissDate in Keeper
Added HttpStatus in ApiExceptionsHandler
Completed Unit tests

Solykk committed:
#9 Completed Unit tests
Completed Integration tests
Completed Acceptance tests

VadimDyachenko commented on src/main/java/juja/microservices/keepers/entity/Keeper.java:
```diff
@@ -0,0 +1,58 @@
+    private String direction;
+    private String startDate;
+    private String dismissDate;
+    private boolean isActive;
```
У полей startDate и dismissDate должен быть тип Date, а не String.

Solykk replied:
Почему Date? в геймификации String. + с датами в базе всегда проблемы.

VadimDyachenko replied:
Проверь еще раз геймификацию, ветку develop или release.  Проблемы начнутся, когда нужно будет в стрингах найти киперов назначенных в понедельник, такого-то месяца высокостного года. )))

VadimDyachenko commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -1,16 +1,30 @@
+    public String addKeeper(KeeperRequest keeperRequest){
+        if(keepersRepository.findOneByUUId(keeperRequest.getFrom()) == null){
+            throw new UnsupportedKeeperException("Only the keeper can appoint another keeper");
+        }
```
Думаю, что нужно вводить более конкретизированные исключения для каждого метода сервиса и обрабатывать его в хендлере. Здесь мы проверяем правило бизнес логики (только Хранитель может добавить нового хранителя)  соответственно можем кинуть определенное исключение с нужным сообщением и кодом ошибки. Имя, как вариант, может быть AddKeeperException, а UnsupportedKeeperException оставить для других, не поддерживаемых исключений. Для примера смотри как сделано в геймификации.

Solykk replied:
Ок переименую. В хендлере обработал. Все делал по примеру геймификации.

VadimDyachenko commented on src/main/java/juja/microservices/keepers/dao/KeepersRepository.java:
```diff
@@ -1,18 +1,48 @@
+        Date currentDate = new Date(System.currentTimeMillis());
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
+        return dateFormat.format(currentDate);
+    }
```
При работе со временем рекомендую использовать библиотеку java.time (SE8) https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html

Solykk replied:
Опять, так написано в геймификации. Я все делал по примеру.

VadimDyachenko replied:
Смотри, пожалуйста, не ветку мастер она уже устарела, а последнюю актуальную develop или release используется java.time

VadimDyachenko commented on src/main/java/juja/microservices/keepers/dao/KeepersRepository.java:
```diff
@@ -1,18 +1,48 @@
+                                    keeperRequest.getUuid(),
+                                    keeperRequest.getDirection(),
+                                    getFormattedCurrentDate());
+        mongoTemplate.save(keeper);
```
Поле *boolean isActive* в базу пишется? Возможно мы от него избавимся, но пока должны сохранять.

Solykk replied:
Пишется, автоматом. При создании кипера поле устанавливается true, при установке даты приостановки устанавливается false

VadimDyachenko commented on src/main/java/juja/microservices/keepers/exception/ApiErrorStatus.java:
```diff
@@ -63,6 +63,11 @@
             "The exception is Spring exceptions: NoHandlerFoundException"
+    ),
+    UNSUPPORTED_KEEPERS_EXCEPTION(
+            "KPR_D11",
```
Здесь вписывается код конкретно того условия, которое не выполняется, например: KPR-F1-D4

Solykk replied:
Да, перепутал названия.

VadimDyachenko commented on src/main/java/juja/microservices/keepers/exception/ApiErrorStatus.java:
```diff
@@ -63,6 +63,11 @@
             "404 response",
             "The exception is Spring exceptions: NoHandlerFoundException"
+    ),
+    UNSUPPORTED_KEEPERS_EXCEPTION(
```
Имя исключения должно соответствовать методу/бизнесправилу, в котором оно возникает. Так будет информативнее и обработчик ошибок сформирует корректный ответ с соответствующим сообщением. Смотри, как сделано в геймификации.

Solykk replied:
ОК

VadimDyachenko commented on src/main/java/juja/microservices/keepers/exception/ApiExceptionsHandler.java:
```diff
@@ -22,6 +22,8 @@
 import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+import springfox.documentation.spring.web.json.Json;
+import springfox.documentation.spring.web.json.JsonSerializer;
 
```
Не используемые импорты. Удаляй

VadimDyachenko commented on src/test/java/juja/microservices/keepers/entity/KeeperTest.java:
```diff
@@ -0,0 +1,71 @@
+                .concat("}");
+        Assert.assertEquals(expected, keeper.toString());
+    }
+
```
Мне кажется это лишний тест

Solykk replied:
Может быть.

adaleorion replied:
в toString есть какая-то особая логика ?

Solykk replied:
нет, просто по примеру делал.

VadimDyachenko reviewed (commented):
Оставил комментарии

Solykk committed:
#9 Rework after review.
Make Date instead String in Keeper.class
Rename UnsupportedKeeperException to AddKeeperExceptions
Remade everything with what questions arose

VadimDyachenko commented on src/main/java/juja/microservices/keepers/dao/KeepersRepository.java:
```diff
@@ -38,11 +40,5 @@ public Keeper findOneByUUId(String id){
         return mongoTemplate.findOne(new Query(Criteria.where("uuid").is(id)), Keeper.class);
```
Имя переменной id, думаю, следует переименовать в uuid, чтобы соответствовало действительности. Мы же в базе не по id ищем кипера, а по его uuid.

Solykk committed:
#9 Rename variable "id" to "uuid" in KeeperRepository  findOneByUUId(String uuid)
Rename method in ApiExceptionsHandler
Add swagger in KeepersController

adaleorion commented on src/test/java/juja/microservices/keepers/dao/KeepersRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+ * @author Dmitriy Lyashenko
+ */
+
+public class KeepersRepositoryTest {
```
Привет, эти тесты как-то странно выглядят
ты тестируешь моки ?

в этом случае нужно либо замокать  mongotemplate

либо сделать с помощью fakemongo

OlgaKulikova replied:
И все равно тесты вышли странными. Ты моку говоришь ничего не делать и потом проверяешь, что он ничего не сделал. Попробуй все-таки применить в этом классе FakeMongo

adaleorion commented on src/test/java/juja/microservices/keepers/controller/KeepersControllerTest.java:
```diff
@@ -1,32 +1,53 @@
 @RunWith(SpringRunner.class)
 @WebMvcTest(KeepersController.class)
 public class KeepersControllerTest {
+
```
в этих тестах нужно использовать 
такой подход 

https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-testing.html#boot-features-testing-spring-boot-applications-testing-autoconfigured-mvc-tests

т.е. замокать все сервисы и проверить контроллер в изоляции. 

отличие от интеграционных в том что в них нет моков прямых зависимостей

VadimDyachenko commented on src/main/java/juja/microservices/keepers/exception/ApiErrorStatus.java:
```diff
@@ -63,6 +63,11 @@
             "The exception is Spring exceptions: NoHandlerFoundException"
+    ),
+    ADD_KEEPERS_EXCEPTION(
+            "KPR-F1-D4",
```
Подыми вверх этот блок кода, чтобы наши исключения были сгруппированы в одном месте.

VadimDyachenko commented on src/test/java/juja/microservices/acceptance/KeepersAT.java:
```diff
@@ -0,0 +1,64 @@
+
+@RunWith(SpringRunner.class)
+public class KeepersAT extends BaseAcceptanceTest{
+
```
Имя класса договорились писать без сокращений

VadimDyachenko commented on src/test/java/juja/microservices/integration/KeepersControllerIT.java:
```diff
@@ -0,0 +1,95 @@
+
+@RunWith(SpringRunner.class)
+public class KeepersControllerIT extends BaseIntegrationTest{
+
```
Имя класса договорились писать без сокращений

VadimDyachenko commented on src/test/java/juja/microservices/integration/KeepersServiceIT.java:
```diff
@@ -0,0 +1,58 @@
+
+@RunWith(SpringRunner.class)
+public class KeepersServiceIT extends BaseIntegrationTest {
+
```
Имя класса договорились писать без сокращений

VadimDyachenko reviewed (commented):
Убрать сокращения в именах тестов. Поднять блок кода в классе ApiErrorStatus.

Solykk committed:
#9 Renamed abbreviations in tests

adaleorion commented on src/main/java/juja/microservices/keepers/entity/Keeper.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public String toString(){
```
почему отказался от  lombok для toString ?

Solykk committed:
#9 Raised up ADD_KEEPER_EXCEPTIONS in ApiErrorStatus

adaleorion commented on src/main/java/juja/microservices/keepers/entity/Keeper.java:
```diff
@@ -0,0 +1,73 @@
+        this.isActive = true;
+    }
+
+    public void setDismissDate(LocalDateTime dismissDate) {
```
может назвать метод просто dismiss, не расскрывая устройства внутри, 
это очень поможет если вдруг для увольнения нужно будет несколько полей поменять,

OlgaKulikova replied:
это же я так понимаю просто сеттер на поле дата увольнения. Тут конкретика к этому полю, потому считаю это правильным.

adaleorion replied:
договорились :)

adaleorion commented on src/main/java/juja/microservices/keepers/exception/AddKeeperException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * @author Dmitriy Lyashenko
+ */
+public class AddKeeperException extends KeepersException {
```
может лучше назвать по имени ошибки который символизирует этот exception ?

OlgaKulikova replied:
да. Класс не называется с глаголом (ну, как правило). В названии должна звучать причина ошибки.

adaleorion commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -1,16 +1,30 @@
+
+    public String addKeeper(KeeperRequest keeperRequest){
+        if(keepersRepository.findOneByUUId(keeperRequest.getFrom()) == null){
+            throw new AddKeeperException("Only the keeper can appoint another keeper");
```
Надо переименовать exception на более конкретный, 
из его имени не понятно, что он отвечает за проверку прав

adaleorion commented on src/test/java/juja/microservices/integration/KeepersControllerIntegrationTest.java:
```diff
@@ -34,14 +38,58 @@ public void setup() {
-    @UsingDataSet(locations = "/datasets/initEmptyDb.json")
-    public void addKeeperTest() throws Exception {
+    @UsingDataSet(locations = "/datasets/oneKeeperInDB.json")
+    public void addKeeperNotExistUUID() throws Exception {
```
В этом тесте ты фактически проверяешь работу exception handler, надо проверить чтобы конвертация ошибки прошла успешно.

adaleorion replied:
теперь сойдет :)

adaleorion commented on src/test/java/juja/microservices/keepers/controller/KeepersControllerTest.java:
```diff
@@ -43,4 +64,29 @@ private void checkBadRequest(String uri, String jsonContentRequest) throws Excep
+        ResponseEntity<?> expected = new ResponseEntity<>(Collections.singletonList("SomeId"), HttpStatus.OK);
+
+        //When
+        ResponseEntity<?> result = keepersController.addKeeper(keeperRequest);
```
при таком варианте ты не проверяешь  конвертацию результата в json

для полноценной проверки контроллера используй mockMvc;

adaleorion commented on src/test/java/juja/microservices/keepers/controller/KeepersControllerTest.java:
```diff
@@ -1,32 +1,53 @@
 
+    @Inject
+    @InjectMocks
+    KeepersController keepersController = new KeepersController();
```
если ты делаешь     @Inject
тогда зачем создаешь его с помощью new ?

adaleorion commented on src/test/java/juja/microservices/keepers/service/KeepersServiceTest.java:
```diff
@@ -0,0 +1,69 @@
+    private KeepersRepository repository;
+
+    @Test(expected = AddKeeperException.class)
+    public void addKeeperNotExist(){
```
тест не очень удачно назван, он говорит что мы проверяем добавление нового кипера когда его нет

а в реальности насколько я понял мы проверяем ситуацию, когда только кипер может добавить нового кипера.

adaleorion reviewed (commented):
i have some comments, check pls

danilkuznetsov commented:
еще нужно добавить логирование, перед тем как делать merge

adaleorion reviewed (changes requested):
еще нужно добавить логирование, перед тем как делать merge

adaleorion commented on src/main/java/juja/microservices/keepers/entity/Keeper.java:
```diff
@@ -0,0 +1,73 @@
+    private Date startDate;
+
+    @JsonFormat(shape= JsonFormat.Shape.STRING, pattern="yyyy-MM-dd")
+    private Date dismissDate;
```
может установим это поле в дату 01,01,1970 например, чтобы не было null ?

OlgaKulikova replied:
так тут же идет речь о том, что если это поле не будет иметь значения, то по нему не будет записи в базу. А потом соответственно мы можем проверять всех активных киперов, у которых этого поля нет в базе.

adaleorion replied:
договорились :)

OlgaKulikova reviewed (changes requested):
Please look at my comments

Solykk committed:
#9 Renamed AddKeeperException to KeeperAccessException
Replaced toString to @toString lombok in Keeper.class
Corrected tests

Solykk committed:
#9 Added Logging

Solykk commented:
Готово, проверьте.

OlgaKulikova commented on src/main/java/juja/microservices/keepers/controller/KeepersController.java:
```diff
@@ -15,6 +19,7 @@
 import java.net.HttpURLConnection;
+
 import java.util.Collections;
 import java.util.List;
```
очень много пробелов между импортами. Как правило Идея сама хорошо их группирует и пробелов между группами минимум.

OlgaKulikova commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -15,14 +18,18 @@
     public String addKeeper(KeeperRequest keeperRequest){
         if(keepersRepository.findOneByUUId(keeperRequest.getFrom()) == null){
-            throw new AddKeeperException("Only the keeper can appoint another keeper");
+            logger.warn("User '{}' tried to add new 'Keeper' but he is no Keeper", keeperRequest.getFrom());
```
"he is not a keeper"

OlgaKulikova commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -15,14 +18,18 @@
+            throw new KeeperAccessException("Only the keeper can appoint another keeper");
         }
 
+        logger.info("Added new 'Keeper' with uuid {}, from user '{}'", keeperRequest.getUuid(), keeperRequest.getFrom());
```
выходит, в лог у тебя пойдет запись о том, что добавлен новый кипер еще до того, как сохранится в базу - могут быть недостоверные логи. Чтобы это поправить, то можно либо результат сейва сохранить в переменную, потом лог, потом ретерн той переменной. Либо логом отработать на уровень выше - в контроллере.

OlgaKulikova replied:
а почему бы тебе не показать результат в лог. А то ты сообщаешь, что кипер добавлен, но реально не знаешь, есть айдишка или нет.

OlgaKulikova commented on src/test/java/juja/microservices/integration/KeepersControllerIntegrationTest.java:
```diff
@@ -41,18 +42,27 @@ public void setup() {
         //Then
-        mockMvc.perform(post("/keepers")
+        Assert.assertEquals( expected, mockMvc.perform(post("/keepers")
                 .contentType(MediaType.APPLICATION_JSON)
```
я бы вынесла в переменную резалт всего mockMvc.perform, а потом уже сверяла ассертом. Так читабельней.

OlgaKulikova commented on src/test/java/juja/microservices/keepers/controller/KeepersControllerTest.java:
```diff
@@ -66,27 +63,35 @@ private void checkBadRequest(String uri, String jsonContentRequest) throws Excep
 
         //Then
-        Assert.assertEquals(expected, result);
+        Assert.assertEquals("[\"SomeId\"]", result);
```
ассерты можно вынести в статический импорт, чтобы каждый раз не вызывать класс.

OlgaKulikova reviewed (commented):
Please look at my comments and do changes

OlgaKulikova commented:
@Solykk Дима, обрати внимание, что Травис не пропускает твои изменения, значит не проходят тесты. Также не забывай, что может кто-то уже запушил изменения в девелоп и тебе перед последним пушем надо сделать fetch upstream, merge (поправить, если есть конфликты) и затем уже запушить свои изменения.

Solykk committed:
#9 Corrected tests
Corrected logs

Solykk committed:
#9 fot Travis

OlgaKulikova commented on src/test/java/juja/microservices/integration/KeepersControllerIntegrationTest.java:
```diff
@@ -55,14 +55,16 @@ public void addKeeperNotExistUUID() throws Exception {
+                .andReturn().getResponse().getContentAsString();
+
+        //Then
+        Assert.assertEquals(expected, result);
```
а тут статический импорт Assert?

OlgaKulikova reviewed (commented):
look at new comments

OlgaKulikova commented:
@Solykk апдейтни свой бранч

Solykk committed:
#9 Corrected tests

Solykk committed:
#9 Cleared the gaps in the imports

Solykk committed:
Merge remote-tracking branch 'upstream/develop' into KPR-F1

Solykk committed:
#9 Cleared KeeperTest set Zone not default

adaleorion commented on src/main/java/juja/microservices/keepers/controller/KeepersController.java:
```diff
@@ -1,28 +1,47 @@
     public ResponseEntity<?> addKeeper(@Valid @RequestBody KeeperRequest request) {
-        //TODO Should be implemented feature KPR-F1
-        List<String> ids = new ArrayList<>();
+        String keeperId = keepersService.addKeeper(request);
```
Добавь еще логирование на уровне debug c входящим KeeperRequest

adaleorion commented on src/main/java/juja/microservices/keepers/dao/KeepersRepository.java:
```diff
@@ -1,18 +1,39 @@
  */
+
 @Repository
 public class KeepersRepository {
```
Добавь логирование на уровне debug на все методы
какие параметры приходят,
что сохраняется/достается в (из) базы.

adaleorion commented on src/main/java/juja/microservices/keepers/exception/ApiExceptionsHandler.java:
```diff
@@ -54,6 +53,16 @@
+                        .httpStatus(HttpStatus.BAD_REQUEST.value())
+                        .exceptionMessage(ex.getMessage())
+                        .build();
+        return new ResponseEntity<>(message,HttpStatus.BAD_REQUEST);
```
нужно добавить логирование на уровне warn, в этом случае мы обрабатываем exception

adaleorion commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -1,16 +1,35 @@
     private KeepersRepository keepersRepository;
+
+    public String addKeeper(KeeperRequest keeperRequest){
+        if(keepersRepository.findOneByUUId(keeperRequest.getFrom()) == null){
```
здесь аналогично нужно добавить логирование с полным выводом на уровне debug

adaleorion reviewed (changes requested):
добавь логирование, а так не плохо :)

OlgaKulikova commented:
@Solykk Удали пока вообще те тесты, которые не проходят. Замерджим твой пул, потом еще поиграешься с тайм зоун.

Solykk committed:
#9 Cleared KeeperTest

Solykk committed:
#9 Added logger

Solykk committed:
Merge remote-tracking branch 'upstream/develop' into KPR-F1

# Conflicts:
#	src/main/java/juja/microservices/keepers/dao/KeepersRepository.java
#	src/main/java/juja/microservices/keepers/entity/Keeper.java
#	src/main/java/juja/microservices/keepers/service/KeepersService.java

Solykk committed:
#9 Pull/merge

OlgaKulikova commented on src/main/java/juja/microservices/keepers/dao/KeepersRepository.java:
```diff
@@ -36,4 +54,6 @@
         return outMap;
     }
+
+    //TODO Should be implemented work with MongoDB
```
Вот эта тудушка не понятно, откуда осталась, удалить.

OlgaKulikova commented on src/main/java/juja/microservices/keepers/entity/Keeper.java:
```diff
@@ -1,33 +1,51 @@
- * @author Dmitriy Roy
+ * @author Dmitriy Lyashenko
  */
-@Data
```
аннотация @Data объединяет в себе сеттеры, геттеры, тустринги. Почему бы нет?

OlgaKulikova replied:
или потому, что нам нужен отдельный сеттер на дату? тогда это оправдано.

OlgaKulikova commented on src/main/java/juja/microservices/keepers/entity/Keeper.java:
```diff
@@ -1,33 +1,51 @@
+ * @author Dmitriy Lyashenko
  */
-@Data
-@AllArgsConstructor
```
Есть причина, почему не использовать AllargsConstructor, а писать отдельный конструктор?

OlgaKulikova commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -1,22 +1,46 @@
         return keepersRepository.getAllActiveKeepers();
     }
+
+    //TODO Should be implemented internal service logic
```
убрать ненужную тудушку

Solykk committed:
#9 drop TODO

Solykk committed:
#9 added new logic about verification (Keeper with uuid  already keep direction  and he is active)

Solykk committed:
#9 added tests

OlgaKulikova commented on src/main/java/juja/microservices/keepers/dao/KeepersRepository.java:
```diff
@@ -37,6 +37,12 @@ public Keeper findOneByUUId(String uuid){
         return mongoTemplate.findOne(new Query(Criteria.where("uuid").is(uuid)), Keeper.class);
     }
 
+    public Keeper findOneByUUIdAndDirectionIsActive(String uuid, String direction){
```
Можно чуть уменьшить название метода, при этом не потеряв сути - findByUuidIsActive

OlgaKulikova commented on src/main/java/juja/microservices/keepers/exception/ApiErrorStatus.java:
```diff
@@ -10,12 +10,18 @@
 
+    KEEPER_DIRECTION_ACTIVE_EXCEPTION(
+            "KPR-F1-D4",
+            "Sorry, but keeper with the requested uuid already keep the requested direction and he is active",
```
"already keeps"

OlgaKulikova commented on src/main/java/juja/microservices/keepers/exception/KeeperDirectionActiveException.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * @author Dmitriy Lyashenko
+ */
+
```
вот этой строки не должно быть между джава доком и сигнатурой класса

OlgaKulikova commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -31,6 +32,13 @@ public String addKeeper(KeeperRequest keeperRequest){
         }
 
+        if(keepersRepository.findOneByUUIdAndDirectionIsActive(keeperRequest.getUuid(), keeperRequest.getDirection()) != null){
+            logger.warn("Keeper with uuid '{}' already keep direction '{}' and he is active",
```
"already keeps"

OlgaKulikova commented on src/main/java/juja/microservices/keepers/service/KeepersService.java:
```diff
@@ -31,6 +32,13 @@ public String addKeeper(KeeperRequest keeperRequest){
+        if(keepersRepository.findOneByUUIdAndDirectionIsActive(keeperRequest.getUuid(), keeperRequest.getDirection()) != null){
+            logger.warn("Keeper with uuid '{}' already keep direction '{}' and he is active",
+                    keeperRequest.getUuid(), keeperRequest.getDirection());
+            throw new KeeperDirectionActiveException("Keeper with uuid " + keeperRequest.getUuid() + " already keep direction "
```
"already keeps"

Solykk committed:
#9 'already keeps'
